### PR TITLE
src: scheduler worker pool

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Split operations on rocksdb from scheduler to worker pool.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@siddontang @ngaut @BusyJay @disksing @shenli PTAL

File: src/storage/txn/latch.rs [hunk 1]
```diff
@@ -143,10 +144,6 @@ mod tests {
         let mut acquired_b = latches.acquire(&mut lock_b, cid_b);
         assert_eq!(acquired_b, false);
 
-        // b retry acquire lock, failed again
```
Comment (BusyJay) [inline src/storage/txn/latch.rs]: Why remove this?

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
-                                Ok(Some(x)) => res.push(Ok((k.raw().unwrap(), x))),
-                                Ok(None) => {}
-                                Err(e) => res.push(Err(::storage::Error::from(e))),
-                            }
+        Command::BatchGet { ref keys, start_ts, .. } => {
+            let snap_store = SnapshotStore::new(snapshot.as_ref(), start_ts);
+            match snap_store.batch_get(keys) {
+                Ok(results) => {
+                    let mut res = vec![];
+                    for (k, v) in keys.into_iter().zip(results.into_iter()) {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Just `results` is enough.

File: src/storage/mod.rs [hunk 3]
```diff
@@ -57,66 +57,65 @@ impl Mutation {
 
 use kvproto::kvrpcpb::Context;
 
+pub enum CallbackType {
```
Comment (BusyJay) [inline src/storage/mod.rs]: `StorageCallback`, `StorageCb` or `KvCb` seem better.

File: src/storage/txn/scheduler.rs [hunk 4]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
+        }
+    }
+}
 
-                let pr: ProcessResult = ProcessResult::Nothing;
-                let cb = make_write_cb(pr, cid, self.schedch.clone());
-                try!(self.engine.async_write(ctx, txn.modifies(), cb));
+fn process_write_impl(cid: u64, cmd: Command, ch: SendCh<Msg>, snapshot: &Snapshot) -> Result<()> {
+    box_try!(match cmd {
+        Command::Prewrite { ref mutations, ref primary, start_ts, .. } => {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Why use ref here?

File: src/storage/txn/scheduler.rs [hunk 5]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
+            _ => Lock::new(vec![]),
+        }
+    }
+
+    fn process_by_worker(&mut self, cid: u64, snapshot: Box<Snapshot>) {
+        debug!("process cmd with snapshot, cid={}", cid);
+        let cmd = {
+            let ctx = self.cmd_ctxs.get(&cid).unwrap();
+            assert_eq!(ctx.cid, cid);
+            ctx.cmd.clone()
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Why clone？

File: src/storage/txn/scheduler.rs [hunk 6]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
-                let mut results = vec![];
-                for m in mutations {
-                    match txn.prewrite(m.clone(), primary) {
-                        Ok(_) => results.push(Ok(())),
-                        e @ Err(MvccError::KeyIsLocked { .. }) => {
-                            results.push(e.map_err(Error::from))
+        }
+        Command::Scan { ref start_key, limit, start_ts, .. } => {
+            let snap_store = SnapshotStore::new(snapshot.as_ref(), start_ts);
+            let scanner = snap_store.scanner();
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `snap_store.scanner().and_then(|mut scanner| {})`

File: src/storage/txn/scheduler.rs [hunk 7]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
                 }
+                Err(e) => ProcessResult::Failed { err: StorageError::from(e) },
+            }
+        }
+        _ => panic!("unsupported read command"),
+    };
 
-                let pr: ProcessResult = ProcessResult::ResultSet { result: results };
-                let cb = make_write_cb(pr, cid, self.schedch.clone());
+    if let Err(e) = ch.send(Msg::ReadFinished { cid: cid, pr: pr }) {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: If this happen, some data may stay in memory for ever. Should add a todo to clean it up.

File: src/storage/txn/latch.rs (hunk 1 above)
Comment (zhangjinpeng1987) [inline src/storage/txn/latch.rs]: Can not retry, cause we have removed TreeSet which check exist.

File: src/storage/txn/scheduler.rs [hunk 8]
```diff
@@ -13,76 +13,133 @@
...
+    SnapshotFinished {
         cid: u64,
         snapshot: EngineResult<Box<Snapshot>>,
     },
-    WriteFinish {
+    ReadFinished {
+        cid: u64,
+        pr: ProcessResult,
+    },
+    PrepareWriteFinished {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `WriteFinished` is better.

File: src/storage/txn/scheduler.rs [hunk 9]
```diff
@@ -345,101 +445,87 @@ impl Scheduler {
...
-                    }
-                };
-                if finished {
-                    self.finish_cmd(cid);
-                }
-            }
-            Err(e) => {
-                self.finish_with_err(cid, Error::from(e));
-            }
+    fn on_read_finished(&mut self, cid: u64, pr: ProcessResult) {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: Can this be merged with `on_prepare_write_failed`?

File: src/storage/txn/scheduler.rs [hunk 10]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
-            }
-            Command::Commit { ref ctx, ref keys, lock_ts, commit_ts, .. } => {
-                let mut txn = MvccTxn::new(snapshot, lock_ts);
-                for k in keys {
-                    try!(txn.commit(&k, commit_ts));
-                }
+fn process_write(cid: u64, cmd: Command, ch: SendCh<Msg>, snapshot: Box<Snapshot>) {
+    if let Err(e) = process_write_impl(cid, cmd, ch.clone(), snapshot.as_ref()) {
+        if let Err(err) = ch.send(Msg::PrepareWriteFailed { cid: cid, err: e }) {
+            error!("send PrepareWriteFailed message to channel failed. cid={}, err={:?}",
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: If this failed, there may be deadlock, maybe panic.

File: src/storage/txn/scheduler.rs [hunk 11]
```diff
@@ -112,177 +172,214 @@ impl Scheduler {
...
+                    let key = start_key.clone();
+                    match scanner.scan(key, limit) {
+                        Ok(mut results) => {
+                            let pairs = results.drain(..)
+                                .map(|x| x.map_err(StorageError::from))
+                                .collect();
+                            ProcessResult::MultiKvpairs { pairs: pairs }
                         }
-                        Err(e) => return Err(Error::from(e)),
+                        Err(e) => ProcessResult::Failed { err: StorageError::from(e) },
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `e.into`

File: src/storage/txn/scheduler.rs [hunk 12]
```diff
@@ -345,101 +445,87 @@ impl Scheduler {
...
+    }
+
+    fn on_prepare_write_finished(&mut self,
+                                 cid: u64,
+                                 pr: ProcessResult,
+                                 to_be_write: Vec<Modify>) {
+        let res = {
+            let ctx = self.cmd_ctxs.get(&cid).unwrap();
+            assert_eq!(ctx.cid, cid);
+            match ctx.cmd {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `extract_context` is enough.

File: src/storage/txn/scheduler.rs [hunk 13]
```diff
@@ -345,101 +445,87 @@ impl Scheduler {
...
+                Command::RollbackThenGet { ref ctx, .. } => {
+                    let engine_cb = make_engine_cb(cid, pr, self.schedch.clone());
+                    self.engine.async_write(ctx, to_be_write, engine_cb)
+                }
+                _ => panic!("unsupported write command"),
             }
+        };
+        if let Err(e) = res {
+            let mut ctx = self.cmd_ctxs.remove(&cid).unwrap();
+            assert_eq!(ctx.cid, cid);
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: I think L489 did it before.

File: src/storage/txn/scheduler.rs (hunk 8 above)
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Should distinguish from WriteFinished.

**Comment by siddontang** [comment]
Any benchmark result?

File: src/storage/mod.rs [hunk 14]
```diff
@@ -57,66 +57,65 @@ impl Mutation {
 
 use kvproto::kvrpcpb::Context;
 
+pub enum StorageCb {
+    Result(Callback<()>),
```
Comment (siddontang) [inline src/storage/mod.rs]: Result and following MultiResults confuse me, what's result for? does Value and KvPairs are not result?

**Comment by zhangjinpeng1987** [comment]
@disksing @ngaut @siddontang @BusyJay PTAL

File: src/storage/txn/scheduler.rs [hunk 15]
```diff
@@ -34,55 +34,111 @@ pub enum Tick {
...
-    WriteFinish {
+    ReadFinished {
+        cid: u64,
+        pr: ProcessResult,
+    },
+    PrepareWriteFinished {
+        cid: u64,
+        cmd: Command,
+        pr: ProcessResult,
+        to_be_write: Vec<Modify>,
```
Comment (disksing) [inline src/storage/txn/scheduler.rs]: to_be_**written**?

File: src/storage/txn/scheduler.rs [hunk 16]
```diff
@@ -352,9 +351,9 @@ impl Scheduler {
     fn process_by_worker(&mut self, cid: u64, snapshot: Box<Snapshot>) {
         debug!("process cmd with snapshot, cid={}", cid);
         let cmd = {
-            let ctx = self.cmd_ctxs.get(&cid).unwrap();
+            let ctx = &mut self.cmd_ctxs.get_mut(&cid).unwrap();
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: No need `&mut` here.

**Comment by BusyJay** [comment]
rest LGTM

**Comment by disksing** [comment]
LGTM.

**Comment by zhangjinpeng1987** [comment]
In xd's case, shceduler with worker pool performance imporved about 100%.

**Pull request merged**